docs/coding-guidelines.md

```diff
@@ -0,0 +1,269 @@
+# Coding Guidelines
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [Coding Guidelines](#coding-guidelines)
+    - [Hierarchical State Machines](#hierarchical-state-machines)
+        - [Conventions for `poll` implementations](#conventions-for-poll-implementations)
+        - [Prioritize local work over new work from a remote](#prioritize-local-work-over-new-work-from-a-remote)
+    - [Bound everything](#bound-everything)
+        - [Channels](#channels)
+        - [Local queues](#local-queues)
+        - [Further reading](#further-reading)
+    - [No premature optimizations](#no-premature-optimizations)
+    - [Keep things sequential unless proven to be slow](#keep-things-sequential-unless-proven-to-be-slow)
+    - [Use `async/await` for sequential execution only](#use-asyncawait-for-sequential-execution-only)
+    - [Don't communicate by sharing memory; share memory by communicating.](#dont-communicate-by-sharing-memory-share-memory-by-communicating)
+        - [Further Reading](#further-reading)
+
+<!-- markdown-toc end -->
+
+
+Below is a set of coding guidelines followed across the rust-libp2p code base.
+
+## Hierarchical State Machines
+
+If you sqint, rust-libp2p is just a big hierarchy of [state
+machines](https://en.wikipedia.org/wiki/Finite-state_machine) where parents pass
```

Comment:
Personally, I am a fan of using a "one-line per sentence" rule for version-controlled markdown documents.

Git works on a line-by-line basis and sentences in natural language tend to change as a whole too. Using one sentence per line makes it really easy to use GitHub's suggestions to collaborate on documents like this and diffs are more meaningful too.

Just an idea :)

Reply:
But I have to admit I am also in the "use an editor with word wrap enabled"-camp :D

Reply:
@mxinden Any comment on this? We don't have to resolve it here. Should I open a separate discussion?

Reply:
Like this, correct @thomaseizinger? https://github.com/mxinden/rust-libp2p/commit/ecf86c81dd54fb6180749131cc6cb630883bfecd

With GitHubs diff highlighting showing what within a line has changed, I don't think there is a great benefit for it. As a downside, I think it adds complexity, e.g. having to explain this to newcomers.

> Just an idea :)

I don't feel strongly about it. It seems like you don't feel strongly about it either. Thus I will proceed without the change.

That said, feel free to propose a new pull request with the commit above.

---

> But I have to admit I am also in the "use an editor with word wrap enabled"-camp :D

Same here. I even stopped truncating my emails to 80 characters :exploding_head:

Reply:
If I get to choose, I'd pick one sentence per line.

We don't have to enforce it though (I am yet to find a tool that reliably does).

Are you happy with doing it on a personal preference basis?

Reply:
> Are you happy with doing it on a personal preference basis?

For markdown files on rust-libp2p? Yes, that is fine by me.